File: 436.find-right-interval.cpp
/*
 * @lc app=leetcode id=436 lang=cpp
 *
 * [436] Find Right Interval
 */
#include <vector>
#include <map>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<int> findRightInterval(vector<vector<int>> &intervals)
    {
        vector<long> search_map;
        int len = intervals.size();
        search_map.reserve(len);
        for (int i = 0; i != len; ++i)
        {
            search_map.push_back(intervals[i][0] * len + i);
        }
        sort(search_map.begin(), search_map.end());
        vector<int> res;
        res.reserve(len);
        for (auto &v : intervals)
        {
            auto iter = lower_bound(search_map.begin(), search_map.end(), v[1] * len);
            if (iter != search_map.end())
            {
                res.push_back((*iter % len + len) % len);
            }
            else
            {
                res.push_back(-1);
            }
        }
        return res;
    }
};
// @lc code=end

File: java/436.find-right-interval.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;

/*
 * @lc app=leetcode id=436 lang=java
 *
 * [436] Find Right Interval
 */

// @lc code=start
class Solution {
    public int[] findRightInterval(int[][] intervals) {
        int[] res = new int[intervals.length];
        ArrayList<Long> search_map = new ArrayList<>(intervals.length);
        for (int i = 0; i < intervals.length; ++i) {
            search_map.add((long) intervals[i][0] * intervals.length + i);
        }
        Collections.sort(search_map);
        for (int i = 0; i < intervals.length; ++i) {
            Integer ind = Collections.binarySearch(search_map, (long) intervals[i][1] * intervals.length);
            if (ind < 0) {
                ind *= -1;
                ind -= 1;
            }
            res[i] = ind != intervals.length
                    ? (int)((search_map.get(ind) % intervals.length) + intervals.length) % intervals.length
                    : -1;
        }
        return res;
    }
}
// @lc code=end
